### PR TITLE
task-04 added DB, sql init script, put product handler

### PR DESCRIPTION
**Pull request by postc0de**
# [Task 04](https://github.com/rolling-scopes-school/nodejs-aws-tasks/blob/main/task4-rds/task.md) 
[Lecture](https://youtu.be/dfD9jvAlaz4) | [QA](https://www.youtube.com/watch?v=7zd3K6zqmmE&feature=youtu.be&ab_channel=RollingScopesSchool)

What was done? 

1. - [x] Task 4.1 is implemented (see script [create-db-products.ts ](https://github.com/bhondu/nodejs-aws-be/pull/2/files#diff-87e66b1f1c7fd3ee1a98ca85fb002a5bffa5c2c6b5c0b554019f420e1b8918ea))
image

3. - [x] TASK 4.2 [`getProducts`](https://ydrurck6qi.execute-api.eu-west-1.amazonaws.com/dev/product) [Swagger API](https://editor.swagger.io/?url=https://raw.githubusercontent.com/bhondu/nodejs-aws-be/bae629da3db1b6cdbb78c07ebc3295b5186f510c/product-service/api.yaml) is implemented lambda links are provided  and returns data

4. - [x] TASK 4.3 [`putProducts`](https://ydrurck6qi.execute-api.eu-west-1.amazonaws.com/dev/product) [Swagger API](https://editor.swagger.io/?url=https://raw.githubusercontent.com/bhondu/nodejs-aws-be/bae629da3db1b6cdbb78c07ebc3295b5186f510c/product-service/api.yaml) is implemented lambda links are provided and products is stored in DB (call TASK 4.2 to see the product)

5. - [x] Your own [Frontend application](https://d1bf2hwgu6wjxf.cloudfront.net/) is integrated with product service (/products API) and products from product-service are represented on Frontend. Link to a working Front-End application is provided for cross-check reviewer.

Additional (optional) tasks:

1. - [x] PUT/products lambda function returns error 400 status code if product data is invalid
image

1. - [x] All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
image

1. - [x] All lambdas do console.log for each incoming requests and their arguments
1. - [x] Transaction based creation of a product. Try [creating a new product](https://d1bf2hwgu6wjxf.cloudfront.net/admin/product-form/) with a negative count. Count update should fail due to constraint violation, the created product is rolled back - no new product can be seen in [`getProducts`](https://ydrurck6qi.execute-api.eu-west-1.amazonaws.com/dev/product).

Link to FE Application: https://d1bf2hwgu6wjxf.cloudfront.net/

### CODE REVIEW

File: product-service/api.yaml [hunk 1]
```diff
@@ -21,8 +22,27 @@ paths:
                 type: array
                 items:
                   $ref: '#/components/schemas/Product'
+    put:
```
Comment (aleksei-bulgak-study) [inline product-service/api.yaml:25]: generally put is used only for update operation and it some kind of best practice to use post for create operation

File: product-service/serverless.ts [hunk 2]
```diff
@@ -26,6 +26,9 @@ const serverlessConfiguration: Serverless = {
     },
     environment: {
       AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
+      PG_HOST: 'nodejs-aws-be.cupqagybo1vs.eu-west-1.rds.amazonaws.com',
```
Comment (aleksei-bulgak-study) [inline product-service/serverless.ts]: Please use env variables. It is absolutely insecure. It is really sad that our lector did not mentioned about this. But I strongly suggest you to replace this hardcoded values with env variables

File: product-service/api.yaml (hunk 1 above)
Comment (postc0de) [reply product-service/api.yaml:25]: Totally agree, just wanted it to be integrated with FE App where PUT is used mistakenly for both creation and modification operations

File: product-service/serverless.ts (hunk 2 above)
Comment (postc0de) [reply product-service/serverless.ts]: Sure, at least I decided not to commit credentials. In real-world application a secret vault should be used instead, maybe this topic will be covered in future lections.

**Comment by aleksei-bulgak-study** [comment]
Screenshot from 2020-11-10 21-15-23
smth wrong with api please take a look at ?column? count field

**Comment by postc0de** [comment]
> smth wrong with api please take a look at ?column? count field
@aleksei-bulgak-study fixed, thank you for the review!

**Pull request merged**